File: couchbase2-ycsb/src/main/java/com/ubb/master/ycsb/benchmark/MongoDBBenchmark.java
package com.ubb.master.ycsb.benchmark;

import com.codahale.metrics.*;
import com.codahale.metrics.jvm.*;
import com.ubb.master.ycsb.iterator.ByteIterator;
import com.ubb.master.ycsb.enums.Status;
import com.ubb.master.ycsb.db.DB;
import com.ubb.master.ycsb.db.MongoDBClient;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.ubb.master.ycsb.workload.CoreWorkload;
import com.ubb.master.ycsb.workload.Workload;
import org.bson.Document;

import java.io.File;
import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicLong;

public class MongoDBBenchmark {
    private static final String HOST = "127.0.0.1";
    private static final String DATABASE_NAME = "ycsb";
    private static final String COLLECTION_NAME = "usertable";
    private static final int RECORD_COUNT = 100000;
    private static final int OPERATION_COUNT = 100000;
    private static final int THREAD_COUNT = 1;

    // Operation proportions
//   private static final double READ_PROPORTION = 0.5;
//    private static final double UPDATE_PROPORTION = 0.5;
//    private static final double READ_PROPORTION = 0.95;
//    private static final double UPDATE_PROPORTION = 0.05;
//    private static final double READ_PROPORTION = 1;
//    private static final double UPDATE_PROPORTION = 0;
    private static final double READ_PROPORTION = 0.05;
    private static final double UPDATE_PROPORTION = 0.95;
    private static final double INSERT_PROPORTION = 0;
    private static final double SCAN_PROPORTION = 0;

    private static final MetricRegistry metrics = new MetricRegistry();
    private static final Counter readCounter = metrics.counter("read.count");
    private static final Counter updateCounter = metrics.counter("update.count");
    private static final Counter insertCounter = metrics.counter("insert.count");
    private static final Counter scanCounter = metrics.counter("scan.count");
    private static final Counter errorCounter = metrics.counter("error.count");

    // Add concurrent lists for storing latencies
    private static final CopyOnWriteArrayList<Long> readLatencies = new CopyOnWriteArrayList<>();
    private static final CopyOnWriteArrayList<Long> updateLatencies = new CopyOnWriteArrayList<>();
    private static final CopyOnWriteArrayList<Long> insertLatencies = new CopyOnWriteArrayList<>();
    private static final CopyOnWriteArrayList<Long> scanLatencies = new CopyOnWriteArrayList<>();

    // Add phase timing variables
    private static final AtomicLong loadPhaseStartTime = new AtomicLong(0);
    private static final AtomicLong loadPhaseEndTime = new AtomicLong(0);
    private static final AtomicLong runPhaseStartTime = new AtomicLong(0);
    private static final AtomicLong runPhaseEndTime = new AtomicLong(0);

    public static void main(String[] args) {
        try {
            // Register JVM metrics
            metrics.register("jvm.gc", new GarbageCollectorMetricSet());
            metrics.register("jvm.memory", new MemoryUsageGaugeSet());
            metrics.register("jvm.threads", new ThreadStatesGaugeSet());
            metrics.register("jvm.files", new FileDescriptorRatioGauge());

            // Initialize MongoDB connection
            MongoClient mongoClient = MongoClients.create("mongodb://" + HOST + ":27017");
            MongoDatabase database = mongoClient.getDatabase(DATABASE_NAME);
            MongoCollection<Document> collection = database.getCollection(COLLECTION_NAME);

            // Create workload
            Properties props = new Properties();
            props.setProperty("recordcount", String.valueOf(RECORD_COUNT));
            props.setProperty("operationcount", String.valueOf(OPERATION_COUNT));
            props.setProperty("fieldcount", "10");
            props.setProperty("fieldlength", "100");
            props.setProperty("readproportion", String.valueOf(READ_PROPORTION));
            props.setProperty("updateproportion", String.valueOf(UPDATE_PROPORTION));
            props.setProperty("scanproportion", String.valueOf(SCAN_PROPORTION));
            props.setProperty("insertproportion", String.valueOf(INSERT_PROPORTION));
            props.setProperty("requestdistribution", "zipfian");

            // MongoDB specific properties
            props.setProperty("mongodb.url", "mongodb://" + HOST + ":27017");
            props.setProperty("mongodb.database", DATABASE_NAME);
            props.setProperty("mongodb.collection", COLLECTION_NAME);
            props.setProperty("mongodb.writeConcern", "ACKNOWLEDGED");
            props.setProperty("mongodb.readPreference", "PRIMARY");
            props.setProperty("mongodb.maxPoolSize", String.valueOf(THREAD_COUNT * 2));
            props.setProperty("mongodb.minPoolSize", String.valueOf(THREAD_COUNT));
            props.setProperty("mongodb.maxIdleTimeMS", "30000");
            props.setProperty("mongodb.maxLifeTimeMS", "300000");
            props.setProperty("mongodb.waitQueueTimeoutMS", "10000");
            props.setProperty("mongodb.connectTimeoutMS", "30000");
            props.setProperty("mongodb.socketTimeoutMS", "10000");

            // Initialize components
            Workload workload = new CoreWorkload();
            workload.init(props);

            DB db = new MongoDBClient();
            db.setProperties(props);
            db.init();

            // Load phase
            System.out.println("Starting load phase...");
            loadData(db, workload, RECORD_COUNT);
            System.out.println("Load phase completed.");

            // Run phase
//            System.out.println("Starting run phase...");
//            runBenchmark(db, workload, OPERATION_COUNT, THREAD_COUNT);
//            System.out.println("Run phase completed.");

            // Save metrics
            saveMetrics();

            // Cleanup
            db.cleanup();
            mongoClient.close();
            workload.cleanup();

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static void loadData(DB db, Workload workload, int recordCount) throws Exception {
        loadPhaseStartTime.set(System.nanoTime());
        
        List<Thread> threads = new ArrayList<>();
        int recordsPerThread = recordCount / THREAD_COUNT;
        
        for (int i = 0; i < THREAD_COUNT; i++) {
            final int threadId = i;
            Thread thread = new Thread(() -> {
                try {
                    int startRecord = threadId * recordsPerThread;
                    int endRecord = (threadId == THREAD_COUNT - 1) ? recordCount : (threadId + 1) * recordsPerThread;
                    
                    for (int j = startRecord; j < endRecord; j++) {
                        String key = String.format("user%d", j);
                        HashMap<String, ByteIterator> values = new HashMap<>();
                        workload.insertInit(key, values);
                        long startTime = System.nanoTime();
                        Status status = db.insert("usertable", key, values);
                        long endTime = System.nanoTime();
                        insertLatencies.add(endTime - startTime);
                        
                        if (status != Status.OK) {
                            System.err.println("Error inserting record: " + key);
                            errorCounter.inc();
                        }
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    errorCounter.inc();
                }
            });
            threads.add(thread);
            thread.start();
        }

        for (Thread thread : threads) {
            thread.join();
        }
        
        loadPhaseEndTime.set(System.nanoTime());
    }

    private static void runBenchmark(DB db, Workload workload, int operationCount, int threadCount) throws Exception {
        runPhaseStartTime.set(System.nanoTime());
        List<Thread> threads = new ArrayList<>();
        for (int i = 0; i < threadCount; i++) {
            Thread thread = new Thread(() -> {
                try {
                    for (int j = 0; j < operationCount / threadCount; j++) {
                        String key = workload.nextTransactionKey();
                        if (key == null) {
                            break;
                        }

                        double random = Math.random();
                        if (random < READ_PROPORTION) {
                            // Perform read operation
                            HashMap<String, ByteIterator> result = new HashMap<>();
                            long startTime = System.nanoTime();
                            Status status = db.read("usertable", key, null, result);
                            long endTime = System.nanoTime();
                            readLatencies.add(endTime - startTime);
                            readCounter.inc();
                            if (status != Status.OK) {
                                errorCounter.inc();
                            }
                        } else if (random < READ_PROPORTION + UPDATE_PROPORTION) {
                            // Perform update operation
                            HashMap<String, ByteIterator> values = new HashMap<>();
                            workload.updateInit(key, values);
                            long startTime = System.nanoTime();
                            Status status = db.update("usertable", key, values);
                            long endTime = System.nanoTime();
                            updateLatencies.add(endTime - startTime);
                            updateCounter.inc();
                            if (status != Status.OK) {
                                errorCounter.inc();
                            }
                        } else if (random < READ_PROPORTION + UPDATE_PROPORTION + INSERT_PROPORTION) {
                            // Perform insert operation
                            HashMap<String, ByteIterator> values = new HashMap<>();
                            workload.insertInit(key, values);
                            long startTime = System.nanoTime();
                            Status status = db.insert("usertable", key, values);
                            long endTime = System.nanoTime();
                            insertLatencies.add(endTime - startTime);
                            insertCounter.inc();
                            if (status != Status.OK) {
                                errorCounter.inc();
                            }
                        } else {
                            // Perform scan operation
                            Vector<HashMap<String, ByteIterator>> result = new Vector<>();
                            long startTime = System.nanoTime();
                            Status status = db.scan("usertable", key, 10, null, result);
                            long endTime = System.nanoTime();
                            scanLatencies.add(endTime - startTime);
                            scanCounter.inc();
                            if (status != Status.OK) {
                                errorCounter.inc();
                            }
                        }
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    errorCounter.inc();
                }
            });
            threads.add(thread);
            thread.start();
        }

        for (Thread thread : threads) {
            thread.join();
        }
        runPhaseEndTime.set(System.nanoTime());
    }

    private static void saveMetrics() throws Exception {
        ObjectMapper mapper = new ObjectMapper();
        Map<String, Object> metricsMap = new HashMap<>();
        
        // Add phase timing metrics
        double loadPhaseTime = (loadPhaseEndTime.get() - loadPhaseStartTime.get()) / 1_000_000.0;
        double runPhaseTime = (runPhaseEndTime.get() - runPhaseStartTime.get()) / 1_000_000.0;
        metricsMap.put("load_phase_time_ms", loadPhaseTime);
        metricsMap.put("run_phase_time_ms", runPhaseTime);
        
        // Add latency metrics for each operation type
        Map<String, Double> readMetrics = calculatePercentiles(readLatencies);
        Map<String, Double> updateMetrics = calculatePercentiles(updateLatencies);
        Map<String, Double> insertMetrics = calculatePercentiles(insertLatencies);
        Map<String, Double> scanMetrics = calculatePercentiles(scanLatencies);
        
        metricsMap.put("read_latencies", readMetrics);
        metricsMap.put("update_latencies", updateMetrics);
        metricsMap.put("insert_latencies", insertMetrics);
        metricsMap.put("scan_latencies", scanMetrics);
        
        // Save to file
        mapper.writerWithDefaultPrettyPrinter()
              .writeValue(new File("metrics.json"), metricsMap);
        
        // Print detailed console output
        System.out.println("\n=== Benchmark Results ===");
        
        // Print phase times
        System.out.println("\nPhase Times:");
        System.out.println("--------------------");
        System.out.printf("Load Phase Time: %.2f ms\n", loadPhaseTime);
        System.out.printf("Run Phase Time:  %.2f ms\n", runPhaseTime);
        
        // Print throughput
        System.out.println("\nThroughput:");
        System.out.println("--------------------");
        System.out.printf("Load Phase: %.2f ops/sec\n", (RECORD_COUNT / (loadPhaseTime / 1000.0)));
        System.out.printf("Run Phase:  %.2f ops/sec\n", (OPERATION_COUNT / (runPhaseTime / 1000.0)));
        
        // Print operation latencies
        System.out.println("\nOperation Latencies (milliseconds):");
        System.out.println("--------------------------------");
        
        // Print read metrics
        if (!readLatencies.isEmpty()) {
            System.out.println("\nRead Operations:");
            System.out.println("--------------------");
            System.out.printf("Count:     %d\n", readLatencies.size());
            System.out.printf("Min:       %.2f ms\n", readMetrics.get("min"));
            System.out.printf("Mean:      %.2f ms\n", readMetrics.get("mean"));
            System.out.printf("P50:       %.2f ms\n", readMetrics.get("p50"));
            System.out.printf("P75:       %.2f ms\n", readMetrics.get("p75"));
            System.out.printf("P90:       %.2f ms\n", readMetrics.get("p90"));
            System.out.printf("P99:       %.2f ms\n", readMetrics.get("p99"));
            System.out.printf("Max:       %.2f ms\n", readMetrics.get("max"));
        }
        
        // Print update metrics
        if (!updateLatencies.isEmpty()) {
            System.out.println("\nUpdate Operations:");
            System.out.println("--------------------");
            System.out.printf("Count:     %d\n", updateLatencies.size());
            System.out.printf("Min:       %.2f ms\n", updateMetrics.get("min"));
            System.out.printf("Mean:      %.2f ms\n", updateMetrics.get("mean"));
            System.out.printf("P50:       %.2f ms\n", updateMetrics.get("p50"));
            System.out.printf("P75:       %.2f ms\n", updateMetrics.get("p75"));
            System.out.printf("P90:       %.2f ms\n", updateMetrics.get("p90"));
            System.out.printf("P99:       %.2f ms\n", updateMetrics.get("p99"));
            System.out.printf("Max:       %.2f ms\n", updateMetrics.get("max"));
        }
        
        // Print insert metrics
        if (!insertLatencies.isEmpty()) {
            System.out.println("\nInsert Operations:");
            System.out.println("--------------------");
            System.out.printf("Count:     %d\n", insertLatencies.size());
            System.out.printf("Min:       %.2f ms\n", insertMetrics.get("min"));
            System.out.printf("Mean:      %.2f ms\n", insertMetrics.get("mean"));
            System.out.printf("P50:       %.2f ms\n", insertMetrics.get("p50"));
            System.out.printf("P75:       %.2f ms\n", insertMetrics.get("p75"));
            System.out.printf("P90:       %.2f ms\n", insertMetrics.get("p90"));
            System.out.printf("P99:       %.2f ms\n", insertMetrics.get("p99"));
            System.out.printf("Max:       %.2f ms\n", insertMetrics.get("max"));
        }
        
        // Print scan metrics
        if (!scanLatencies.isEmpty()) {
            System.out.println("\nScan Operations:");
            System.out.println("--------------------");
            System.out.printf("Count:     %d\n", scanLatencies.size());
            System.out.printf("Min:       %.2f ms\n", scanMetrics.get("min"));
            System.out.printf("Mean:      %.2f ms\n", scanMetrics.get("mean"));
            System.out.printf("P50:       %.2f ms\n", scanMetrics.get("p50"));
            System.out.printf("P75:       %.2f ms\n", scanMetrics.get("p75"));
            System.out.printf("P90:       %.2f ms\n", scanMetrics.get("p90"));
            System.out.printf("P99:       %.2f ms\n", scanMetrics.get("p99"));
            System.out.printf("Max:       %.2f ms\n", scanMetrics.get("max"));
        }
        
        // Print operation distribution
        System.out.println("\nOperation Distribution:");
        System.out.println("--------------------");
        System.out.printf("Read Operations:   %d (%.1f%%)\n", readCounter.getCount(), (readCounter.getCount() * 100.0) / OPERATION_COUNT);
        System.out.printf("Update Operations: %d (%.1f%%)\n", updateCounter.getCount(), (updateCounter.getCount() * 100.0) / OPERATION_COUNT);
        System.out.printf("Insert Operations: %d (%.1f%%)\n", insertCounter.getCount(), (insertCounter.getCount() * 100.0) / OPERATION_COUNT);
        System.out.printf("Scan Operations:   %d (%.1f%%)\n", scanCounter.getCount(), (scanCounter.getCount() * 100.0) / OPERATION_COUNT);
        System.out.printf("Total Errors:      %d\n", errorCounter.getCount());
        
        System.out.println("\nDetailed metrics have been saved to metrics.json");
    }

    private static Map<String, Double> calculatePercentiles(List<Long> latencies) {
        Map<String, Double> percentiles = new HashMap<>();
        if (latencies.isEmpty()) {
            return percentiles;
        }

        // Convert to array and sort
        Long[] sortedLatencies = latencies.toArray(new Long[0]);
        Arrays.sort(sortedLatencies);

        // Calculate percentiles (convert from nanoseconds to milliseconds)
        percentiles.put("p50", sortedLatencies[(int) (sortedLatencies.length * 0.50)] / 1_000_000.0);
        percentiles.put("p75", sortedLatencies[(int) (sortedLatencies.length * 0.75)] / 1_000_000.0);
        percentiles.put("p90", sortedLatencies[(int) (sortedLatencies.length * 0.90)] / 1_000_000.0);
        percentiles.put("p99", sortedLatencies[(int) (sortedLatencies.length * 0.99)] / 1_000_000.0);
        percentiles.put("min", sortedLatencies[0] / 1_000_000.0);
        percentiles.put("max", sortedLatencies[sortedLatencies.length - 1] / 1_000_000.0);
        percentiles.put("mean", latencies.stream().mapToLong(Long::longValue).average().orElse(0) / 1_000_000.0);

        return percentiles;
    }
} 

File: benchmark/mongodb_benchmark.py
import time
import threading
import statistics
from typing import Dict, List, Optional
from concurrent.futures import ThreadPoolExecutor
from pymongo import MongoClient
from pymongo.collection import Collection
from pymongo.database import Database
import json
import os
from datetime import datetime
import psutil
import threading

class MongoDBBenchmark:
    def __init__(self):
        # Configuration constants
        self.HOST = "127.0.0.1"
        self.DATABASE_NAME = "ycsb"
        self.COLLECTION_NAME = "usertable"
        self.RECORD_COUNT = 10000000
        self.OPERATION_COUNT = 10000000
        self.THREAD_COUNTS = [2]  # List of thread counts to test

        # Define workload configurations
        self.WORKLOADS = [
            {
                "name": "Balanced",
                "read_proportion": 0.5,
                "update_proportion": 0.5,
                "insert_proportion": 0,
                "scan_proportion": 0
            },
            {
                "name": "Read-Heavy",
                "read_proportion": 0.95,
                "update_proportion": 0.05,
                "insert_proportion": 0,
                "scan_proportion": 0
            },
            {
                "name": "Read-Only",
                "read_proportion": 1.0,
                "update_proportion": 0.0,
                "insert_proportion": 0,
                "scan_proportion": 0
            },
            {
                "name": "Update-Heavy",
                "read_proportion": 0.05,
                "update_proportion": 0.95,
                "insert_proportion": 0,
                "scan_proportion": 0
            }
        ]

        # Metrics
        self.read_counter = 0
        self.update_counter = 0
        self.insert_counter = 0
        self.scan_counter = 0
        self.error_counter = 0

        # Latency lists
        self.read_latencies: List[float] = []
        self.update_latencies: List[float] = []
        self.insert_latencies: List[float] = []
        self.scan_latencies: List[float] = []

        # Phase timing
        self.load_phase_start_time = 0
        self.load_phase_end_time = 0
        self.run_phase_start_time = 0
        self.run_phase_end_time = 0

        # MongoDB client
        self.client: Optional[MongoClient] = None
        self.db: Optional[Database] = None
        self.collection: Optional[Collection] = None

        # Create metrics directories if they don't exist
        os.makedirs("metrics2", exist_ok=True)
        os.makedirs("load_metrics", exist_ok=True)

    def init_connection(self):
        """Initialize MongoDB connection"""
        self.client = MongoClient(f"mongodb://{self.HOST}:27017?retryWrites=false")
        self.db = self.client[self.DATABASE_NAME]
        self.collection = self.db[self.COLLECTION_NAME]

    def cleanup(self):
        """Cleanup MongoDB connection"""
        if self.client:
            self.client.close()

    def print_phase_metrics(self, phase_name: str, start_time: float, end_time: float, operation_count: int, latencies: List[float]):
        """Print metrics for a specific phase"""
        duration = end_time - start_time
        throughput = operation_count / duration if duration > 0 else 0

        print(f"\n=== {phase_name} Phase Results ===")
        print(f"Duration: {duration:.2f} seconds")
        print(f"Throughput: {throughput:.2f} ops/sec")
        
        if latencies:
            latencies_ms = [lat * 1000 for lat in latencies]  # Convert to milliseconds
            print(f"\nLatency Statistics (ms):")
            print(f"Count: {len(latencies)}")
            print(f"Min: {min(latencies_ms):.2f}")
            print(f"Mean: {statistics.mean(latencies_ms):.2f}")
            print(f"P50: {statistics.median(latencies_ms):.2f}")
            print(f"P75: {sorted(latencies_ms)[int(len(latencies_ms) * 0.75)]:.2f}")
            print(f"P90: {sorted(latencies_ms)[int(len(latencies_ms) * 0.90)]:.2f}")
            print(f"P99: {sorted(latencies_ms)[int(len(latencies_ms) * 0.99)]:.2f}")
            print(f"Max: {max(latencies_ms):.2f}")
        else:
            print("\nNo operations performed in this phase")

    def load_data(self):
        """Load initial data into MongoDB"""
        print("Starting load phase...")
        self.load_phase_start_time = time.time()

        def load_worker(start_record: int, end_record: int):
            for i in range(start_record, end_record):
                key = f"user{i}"
                values = {
                    "field0": f"value{i}_0",
                    "field1": f"value{i}_1",
                    "field2": f"value{i}_2",
                    "field3": f"value{i}_3",
                    "field4": f"value{i}_4",
                    "field5": f"value{i}_5",
                    "field6": f"value{i}_6",
                    "field7": f"value{i}_7",
                    "field8": f"value{i}_8",
                    "field9": f"value{i}_9"
                }

                start_time = time.time()
                try:
                    self.collection.insert_one({"_id": key, **values})
                    self.insert_latencies.append(time.time() - start_time)
                except Exception as e:
                    print(f"Error inserting record {key}: {e}")
                    self.error_counter += 1

        used_threads = 10

        records_per_thread = self.RECORD_COUNT // used_threads
        threads = []

        for i in range(used_threads):
            start_record = i * records_per_thread
            end_record = self.RECORD_COUNT if i == used_threads - 1 else (i + 1) * records_per_thread
            thread = threading.Thread(target=load_worker, args=(start_record, end_record))
            threads.append(thread)
            thread.start()

        for thread in threads:
            thread.join()

        self.load_phase_end_time = time.time()
        print("Load phase completed.")
        
        # Print load phase metrics
        self.print_phase_metrics(
            "Load",
            self.load_phase_start_time,
            self.load_phase_end_time,
            self.RECORD_COUNT,
            self.insert_latencies
        )

        # Save load metrics
        load_metrics = {
            "workload_name": "Load",
            "thread_count": used_threads,
            "record_count": self.RECORD_COUNT,
            "operation_count": self.RECORD_COUNT,
            "load_phase_time": self.load_phase_end_time - self.load_phase_start_time,
            "insert_latencies": self.calculate_percentiles(self.insert_latencies),
            "operation_counts": {
                "insert": self.RECORD_COUNT,
                "errors": self.error_counter
            }
        }

        # Save to file in load_metrics directory
        filename = f"load_metrics/{self.RECORD_COUNT}_mongodb_{used_threads}_Load.json"
        with open(filename, "w") as f:
            json.dump(load_metrics, f, indent=2)

        print(f"\nLoad metrics have been saved to {filename}")

    def run_benchmark(self):
        """Run the benchmark operations for all workloads and thread counts"""
        print("\nStarting benchmark runs for all workloads and thread counts...")
        
        for thread_count in self.THREAD_COUNTS:
            print(f"\n=== Testing with {thread_count} threads ===")
            
            for workload in self.WORKLOADS:
                print(f"\n=== Starting {workload['name']} Workload ===")
                print(f"Read proportion: {workload['read_proportion']}")
                print(f"Update proportion: {workload['update_proportion']}")
                
                # Reset counters and latencies for this workload
                self.read_counter = 0
                self.update_counter = 0
                self.insert_counter = 0
                self.scan_counter = 0
                self.error_counter = 0
                self.read_latencies = []
                self.update_latencies = []
                self.insert_latencies = []
                self.scan_latencies = []
                
                self.run_phase_start_time = time.time()

                def run_worker():
                    for _ in range(self.OPERATION_COUNT // thread_count):
                        key = f"user{hash(str(time.time())) % self.RECORD_COUNT}"
                        random_value = time.time() % 1.0

                        if random_value < workload['read_proportion']:
                            # Read operation
                            start_time = time.time()
                            try:
                                result = self.collection.find_one({"_id": key})
                                self.read_latencies.append(time.time() - start_time)
                                self.read_counter += 1
                            except Exception as e:
                                print(f"Error reading record {key}: {e}")
                                self.error_counter += 1

                        elif random_value < workload['read_proportion'] + workload['update_proportion']:
                            # Update operation
                            values = {
                                "field0": f"updated_value_{time.time()}",
                                "field1": f"updated_value_{time.time()}",
                                "field2": f"updated_value_{time.time()}",
                                "field3": f"updated_value_{time.time()}",
                                "field4": f"updated_value_{time.time()}"
                            }
                            start_time = time.time()
                            try:
                                self.collection.update_one({"_id": key}, {"$set": values})
                                self.update_latencies.append(time.time() - start_time)
                                self.update_counter += 1
                            except Exception as e:
                                print(f"Error updating record {key}: {e}")
                                self.error_counter += 1

                threads = []
                for _ in range(thread_count):
                    thread = threading.Thread(target=run_worker)
                    threads.append(thread)
                    thread.start()

                for thread in threads:
                    thread.join()

                self.run_phase_end_time = time.time()
                print(f"\n{workload['name']} workload completed.")

                # Print run phase metrics for this workload
                print(f"\n=== {workload['name']} Workload Results ===")
                duration = self.run_phase_end_time - self.run_phase_start_time
                total_operations = self.read_counter + self.update_counter
                throughput = total_operations / duration if duration > 0 else 0
                
                print(f"Duration: {duration:.2f} seconds")
                print(f"Total Operations: {total_operations}")
                print(f"Throughput: {throughput:.2f} ops/sec")
                
                if self.read_latencies:
                    print("\nRead Operations:")
                    self.print_phase_metrics("Read", 0, 0, self.read_counter, self.read_latencies)
                else:
                    print("\nNo read operations performed")
                
                if self.update_latencies:
                    print("\nUpdate Operations:")
                    self.print_phase_metrics("Update", 0, 0, self.update_counter, self.update_latencies)
                else:
                    print("\nNo update operations performed")

                # Save metrics for this workload and thread count
                self.save_metrics(workload['name'], thread_count)

    def calculate_percentiles(self, latencies: List[float]) -> Dict[str, float]:
        """Calculate percentile statistics for latencies"""
        if not latencies:
            return {}

        latencies.sort()
        return {
            "min": latencies[0],
            "max": latencies[-1],
            "mean": statistics.mean(latencies),
            "p50": latencies[int(len(latencies) * 0.50)],
            "p75": latencies[int(len(latencies) * 0.75)],
            "p90": latencies[int(len(latencies) * 0.90)],
            "p95": latencies[int(len(latencies) * 0.95)],
            "p99": latencies[int(len(latencies) * 0.99)]
        }

    def save_metrics(self, workload_name: str = "", thread_count: int = 1):
        """Save benchmark metrics to file and print results"""
        metrics = {
            "workload_name": workload_name,
            "thread_count": thread_count,
            "record_count": self.RECORD_COUNT,
            "operation_count": self.OPERATION_COUNT,
            "load_phase_time": self.load_phase_end_time - self.load_phase_start_time,
            "run_phase_time": self.run_phase_end_time - self.run_phase_start_time,
            "read_latencies": self.calculate_percentiles(self.read_latencies),
            "update_latencies": self.calculate_percentiles(self.update_latencies),
            "insert_latencies": self.calculate_percentiles(self.insert_latencies),
            "scan_latencies": self.calculate_percentiles(self.scan_latencies),
            "operation_counts": {
                "read": self.read_counter,
                "update": self.update_counter,
                "insert": self.insert_counter,
                "scan": self.scan_counter,
                "errors": self.error_counter
            }
        }

        # Save to file in metrics directory
        filename = f"metrics2/{self.RECORD_COUNT}_mongodb_{thread_count}_{workload_name}.json"
        with open(filename, "w") as f:
            json.dump(metrics, f, indent=2)

        print(f"\nDetailed metrics have been saved to {filename}")

def main():
    benchmark = MongoDBBenchmark()
    try:
        benchmark.init_connection()
        benchmark.load_data()
        benchmark.run_benchmark()
    finally:
        benchmark.cleanup()

if __name__ == "__main__":
    main() 